### PR TITLE
Java file completion, fallback method completion

### LINKED ISSUE #34: Completion for mp-fault-tolerance fallbackMethod
Would be nice to provide completion fallback method completion.

Ex : fallbackMethod="|" should provide fallback:

### PR DESCRIPTION
**Pull request by datho7561**
Infrastructure to contribute completion to Java files through the JDT component. Completion for fallbackMethod.

Closes #34

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

**Comment by datho7561** [comment]
- [ ] integration tests

**Comment by datho7561** [comment]
- [ ] integration tests

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/java/completion/JavaCompletionDefinition.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.lsp4j.CompletionItem;
+import org.eclipse.lsp4mp.jdt.core.java.completion.IJavaCompletionParticipant;
+import org.eclipse.lsp4mp.jdt.core.java.completion.JavaCompletionContext;
+import org.eclipse.lsp4mp.jdt.internal.core.java.AbstractJavaFeatureDefinition;
+import org.eclipse.lsp4mp.jdt.internal.core.java.codelens.JavaCodeLensDefinition;
+
+/**
+ * Wrap the completion participant
```
Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/java/completion/JavaCompletionDefinition.java:31]: Improve doc

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/ls/MicroProfileDelegateCommandHandlerForJava.java [hunk 2]
```diff
@@ -223,10 +228,56 @@ private static MicroProfileJavaCodeLensParams createMicroProfileJavaCodeLensPara
...
+	 * @throws CoreException
+	 */
+	private static CompletionList getCompletionForJava(List<Object> arguments, String commandId,
+			IProgressMonitor monitor) throws JavaModelException, CoreException {
+		MicroProfileJavaCompletionParams params = createMicroProfileJavaCompletionParams(arguments, commandId);
+		return PropertiesManagerForJava.getInstance().completion(params, JDTUtilsLSImpl.getInstance(), monitor);
+	}
+
+	/**
+	 * Create the completion parameters from the given argument map
```
Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/ls/MicroProfileDelegateCommandHandlerForJava.java:248]: Returns...

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	private static IAnnotation getFallbackAnnotation(ITypeRoot typeRoot, int offset) throws JavaModelException {
+		IJavaElement element = typeRoot.getElementAt(offset);
+		IAnnotation annotation = null;
+		switch (element.getElementType()) {
+		case IJavaElement.ANNOTATION:
+			annotation = (IAnnotation) element;
+			break;
+		case IJavaElement.METHOD:
+			IAnnotatable annotatable = (IAnnotatable) element;
+			annotation = annotatable.getAnnotation("Fallback");
```
Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java:100]: Extract to constant

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import org.eclipse.lsp4mp.jdt.core.utils.JDTTypeUtils;
+
+/**
+ * Completion for <code>fallbackMethod</code>
+ *
+ * @author datho7561
+ */
+public class MicroProfileFaultToleranceCompletionParticipant implements IJavaCompletionParticipant {
+
+	private static final Pattern PTN = Pattern.compile("fallbackMethod\\s*=\\s*\"([^\"]*)\"", Pattern.DOTALL);
```
Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java:53]: Better name

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	 * @param typeRoot
+	 * @param offset
+	 * @return the <code>@Fallback</code> annotation as an IAnnotation or null if the offset is not in a <code>@Fallback</code> annotation
+	 * @throws JavaModelException if this error is thrown when accessing the model
+	 */
+	private static IAnnotation getFallbackAnnotation(ITypeRoot typeRoot, int offset) throws JavaModelException {
+		IJavaElement element = typeRoot.getElementAt(offset);
+		IAnnotation annotation = null;
+		switch (element.getElementType()) {
+		case IJavaElement.ANNOTATION:
```
Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java:95]: Test to see if its possible to reach this case

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+			break;
+		}
+		if (!AnnotationUtils.isMatchAnnotation(annotation, FALLBACK_ANNOTATION)) {
+			return null;
+		}
+		return annotation;
+	}
+
+	/**
+	 *
```
Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java:110]: Complete docs

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+		}
+		List<CompletionItem> completionItems = new ArrayList<>();
+		for (IMethod method : context.getTypeRoot().findPrimaryType().getMethods()) {
+			completionItems.add(makeMethodCompletionItem(method.getElementName(), range));
+		}
+		return completionItems;
+	}
+
+	/**
+	 *
```
Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceCompletionParticipant.java:84]: Complete docs

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/commons/MicroProfileJavaCompletionParams.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lsp4mp.commons;
+
+import org.eclipse.lsp4j.Position;
+
+public class MicroProfileJavaCompletionParams {
```
Comment (datho7561) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/commons/MicroProfileJavaCompletionParams.java:18]: Docs

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/PropertiesManagerForJava.java [hunk 9]
```diff
@@ -156,6 +161,54 @@ private void collectCodeLens(String uri, ITypeRoot typeRoot, IJDTUtils utils, Mi
 		definitions.forEach(definition -> definition.endCodeLens(context, monitor));
 	}
 
+	/**
+	 * Returns the CompletionItems given the completion item params
+	 *
+	 * @param params the completion item params
```
Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/PropertiesManagerForJava.java:167]: Format this

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/MicroProfileFaultToleranceJavaCompletionTest.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+				.getFile(new Path("src/main/java/org/acme/OtherFaultToleranceResource.java"));
+		String javaFileUri = fixURI(javaFile.getLocation().toFile().toURI());
+
+		assertJavaCompletion(new MicroProfileJavaCompletionParams(javaFileUri, p(28, 32)), utils, //
+				c(te(28, 32, 28, 32, "hello"), "hello()", CompletionItemKind.Method), //
+				c(te(28, 32, 28, 32, "hi"), "hi()", CompletionItemKind.Method), //
+				c(te(28, 32, 28, 32, "third"), "third()", CompletionItemKind.Method), //
+				c(te(28, 32, 28, 32, "aaa"), "aaa()", CompletionItemKind.Method));
+	}
+
```
Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/MicroProfileFaultToleranceJavaCompletionTest.java:103]: Case for
```java
@Fallback(fallbackMethod="")
```
(no space around "")

Comment (datho7561) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/MicroProfileFaultToleranceJavaCompletionTest.java:103]: Case for multiline annotation eg.
```java
@Fallback(fallbackMethod=
    "|")
```

**Comment by datho7561** [comment]
Does not work with vscode-java >=0.77.x, due to eclipse.jdt.ls upgrading lsp4j to 0.12.0

**Comment by angelozerr** [comment]
> Does not work with vscode-java >=0.77.x, due to eclipse.jdt.ls upgrading lsp4j to 0.12.0

JDT LS is using 0.11.0 LSP4J (not 0.12.0) but yes it requires some changes with Either to fix problem compilation.

File: microprofile.jdt/pom.xml [hunk 11]
```diff
@@ -14,7 +14,7 @@
     <tycho.extras.version>${tycho.version}</tycho.extras.version>
     <tycho.scmUrl>scm:git:https://github.com/eclipse/lsp4mp</tycho.scmUrl>
     <tycho.generateSourceReferences>true</tycho.generateSourceReferences>
-    <jdt.ls.version>0.57.0.20200619102332</jdt.ls.version>
```
Comment (datho7561) [inline microprofile.jdt/pom.xml:17]: We might want to move this part to a different PR. What was the rational for using 0.57.0?

**Comment by angelozerr** [comment]
There are some perfromance problem when the MicroProfile LS is integrated inside vscode-microprofile (communication performance problem between MicroProfile LS -> vscode-microprofile -> JDT LS).

In Eclipse IDE we will not have this problem, we should finda better solution to fix the problem (ex : desactivate the completion support for vscode-microprofile and provide completion by implementing a typescript completipon provider which consumes directly the JDT LS completion command to by pass the communication between vscode-microprofile -> MP LS -> vscode-microprofile -> JDT LS -> vscode-microprofile -> MP LS -> vscode-microprofile (just having vscode-microprofile -> JDT LS -> vscode-microprofile)

**Comment by datho7561** [comment]
@angelozerr I have added a commit that implements the completion the way you suggested. It contributes the command `microprofile/java/completion` in the jdt.ls extension, which returns the completion results given the document uri and cursor position. I also made an example implementation of a client that consumes this support (redhat-developer/vscode-microprofile#57).

To test the performance, made a class with a jax-rs function that has the `@Fallback` annotation. Then, I use a python script to add 2000 methods to the class. When I open completion at `...fallbackMethod="|"`, it takes under 2 seconds for both implementations. It might be a little faster through the client side implementation, but I would need to conduct more timing tests.

**Comment by datho7561** [comment]
I did some timing tests using these steps:
1. Opened the file I created that I mentioned in the previous
2. Waited for jdt.ls and lsp4mp to start up
3. Opened completion on the fallback method annotation, then closed the completion, added a println to the annotated method, opened completion on the fallback method again, then deleted the println
4. Repeated step 3 a few times to get a larger sample size
5. Did steps 1 - 4 for both implementations 

I found on average, the implementation that added code on the client side to invoke completion took 718ms, and the other implementation took 783ms. This is an 8.35% performance improvement.

I think this strategy is worth pursuing. I'll clean up the code for my implementation of the strategy, and add a setting to toggle between the two strategies so that a client that doesn't support scripting command calls can still use this functionality.

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/settings/MicroProfileCompletionSettings.java [hunk 12]
```diff
@@ -1,59 +1,31 @@
...
-	 * @return <code>true</code> if the client support the given documentation
-	 *         format and <code>false</code> otherwise.
-	 */
-	public boolean isDocumentationFormatSupported(String documentationFormat) {
-		return completionCapabilities != null && completionCapabilities.getCompletionItem() != null
-				&& completionCapabilities.getCompletionItem().getDocumentationFormat() != null
-				&& completionCapabilities.getCompletionItem().getDocumentationFormat().contains(documentationFormat);
-	}
-}
+package org.eclipse.lsp4mp.settings;
```
Comment (datho7561) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/settings/MicroProfileCompletionSettings.java:1]: copyright header

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/settings/MicroProfileCompletionSettings.java [hunk 13]
```diff
@@ -1,59 +1,31 @@
...
+
+public class MicroProfileCompletionSettings {
```
Comment (datho7561) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/settings/MicroProfileCompletionSettings.java:3]: Javadoc

File: microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/java/JavaFileTextDocumentService.java [hunk 14]
```diff
@@ -139,11 +151,28 @@ public void didSave(DidSaveTextDocumentParams params) {
...
+				if (list1 == null) {
+					list1 = new CompletionList();
+					list1.setItems(new ArrayList<>());
+				}
+				if (list2 != null) {
+					for (CompletionItem item : list2.getItems()) {
+						list1.getItems().add(item);
+					}
+				}
+				// This is set to hopefully reduce the number of
```
Comment (datho7561) [inline microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/ls/java/JavaFileTextDocumentService.java:172]: ???

**Comment by datho7561** [comment]
Need to rework to use extended client capabilities instead of a setting

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/commons/MicroProfileJavaCompletionParams.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
+/*******************************************************************************
+* Copyright (c) 2021 Red Hat Inc. and others.
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/commons/MicroProfileJavaCompletionParams.java:2]: Header should be 2022, please update all hearders.

File: microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/java/completion/JavaCompletionDefinition.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+		}
+	}
+
+	@Override
+	public List<? extends CompletionItem> collectCompletionItems(JavaCompletionContext context, IProgressMonitor monitor) {
+		try {
+			List<? extends CompletionItem> completionItems = getParticipant().collectCompletionItems(context, monitor);
+			return completionItems != null ? completionItems : Collections.emptyList();
+		} catch (Exception e) {
+			LOGGER.log(Level.SEVERE, "Error while calling getCompletionItems", e);
```
Comment (angelozerr) [inline microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/core/java/completion/JavaCompletionDefinition.java:62]: "Error while calling collectCompletionItems"

**Comment by angelozerr** [comment]
The code is very impressive, congrats @datho7561 !

Let me play with your PR.

**Comment by angelozerr** [comment]
Great job @datho7561 !
